src/flows/context/flow.current.tsx

```diff
@@ -125,80 +222,119 @@ export const FlowProvider: FC = ({children}) => {
 
-    delete initial.title
-    initial.id = id
+      delete initial.title
+      initial.id = id
+
+      if (isFlagEnabled('sharedFlowEditing')) {
+        const flowCopy = JSON.parse(JSON.stringify(currentFlow))
+        flowCopy.data.byID[id] = initial
+        flowCopy.meta.byID[id] = {
+          title,
+          visible: true,
+        }
+        if (typeof index !== 'undefined') {
+          flowCopy.data.allIDs.splice(index + 1, 0, id)
+          flowCopy.meta.allIDs.splice(index + 1, 0, id)
+        } else {
+          flowCopy.data.allIDs.push(id)
+          flowCopy.meta.allIDs.push(id)
+        }
+        yDoc.current.getMap('localState').set('localState', {
+          ...flowCopy,
+        })
+        return
+      }
+      if (isFlagEnabled('ephemeralNotebook') && !currentFlow.id) {
+        currentFlow.data.byID[id] = initial
+        currentFlow.meta.byID[id] = {
+          title,
+          visible: true,
+        }
+        if (typeof index !== 'undefined') {
+          currentFlow.data.allIDs.splice(index + 1, 0, id)
+          currentFlow.meta.allIDs.splice(index + 1, 0, id)
+        } else {
+          currentFlow.data.allIDs.push(id)
+          currentFlow.meta.allIDs.push(id)
+        }
+        setCurrentFlow({...currentFlow})
+        return
+      }
 
-    if (isFlagEnabled('ephemeralNotebook') && !currentFlow.id) {
-      currentFlow.data.byID[id] = initial
-      currentFlow.meta.byID[id] = {
+      flows[currentID].data.byID[id] = initial
+      flows[currentID].meta.byID[id] = {
         title,
         visible: true,
       }
+
       if (typeof index !== 'undefined') {
-        currentFlow.data.allIDs.splice(index + 1, 0, id)
-        currentFlow.meta.allIDs.splice(index + 1, 0, id)
+        flows[currentID].data.allIDs.splice(index + 1, 0, id)
+        flows[currentID].meta.allIDs.splice(index + 1, 0, id)
       } else {
-        currentFlow.data.allIDs.push(id)
-        currentFlow.meta.allIDs.push(id)
+        flows[currentID].data.allIDs.push(id)
+        flows[currentID].meta.allIDs.push(id)
       }
-      setCurrentFlow({...currentFlow})
-      return
-    }
 
-    flows[currentID].data.byID[id] = initial
-    flows[currentID].meta.byID[id] = {
-      title,
-      visible: true,
-    }
+      updateData(id, {})
+      updateMeta(id, {})
 
-    if (typeof index !== 'undefined') {
-      flows[currentID].data.allIDs.splice(index + 1, 0, id)
-      flows[currentID].meta.allIDs.splice(index + 1, 0, id)
-    } else {
-      flows[currentID].data.allIDs.push(id)
-      flows[currentID].meta.allIDs.push(id)
-    }
+      return id
+    },
+    [currentFlow]
+  )
 
-    updateData(id, {})
-    updateMeta(id, {})
+  const removePipe = useCallback(
+    (id: string) => {
+      if (isFlagEnabled('sharedFlowEditing')) {
+        const flowCopy = JSON.parse(JSON.stringify(currentFlow))
 
-    return id
-  }
+        flowCopy.meta.allIDs = flowCopy.meta.allIDs.filter(_id => _id !== id)
```

Comment:
what's the reason for creating these flow copies everywhere? Can't we just expand the object when we pass it into `.set` like you're doing here with `.set('localState', { ...flowCopy })`?

Reply:
I wish. There was some weird JS equality that was happening where the object that was being referenced was being updated so the dim next triggered a reprender even though the values changed

Reply:
gotchaaa